pkg/elasticetcd/leader.go

```diff
@@ -0,0 +1,269 @@
+package elasticetcd
+
+import (
+	"context"
+	"strconv"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/concurrency"
+	"github.com/coreos/etcd/etcdserver/etcdserverpb"
+	"github.com/coreos/etcd/pkg/types"
+)
+
+const (
+	eePrefix        = "elastic"
+	electionKey     = eePrefix + "/election"
+	volunteerPrefix = eePrefix + "/volunteers/"
+	nomineePrefix   = eePrefix + "/nominees/"
+	idealSizeKey    = eePrefix + "/idealSize"
+)
+
+func (ee *ElasticEtcd) startCampaign() {
+	go func() {
+		election := concurrency.NewElection(ee.session, electionKey)
+		for {
+			// Campaign for the becoming the leader
+			// Stop campaigning if the context is canceled
+			// else ignore errors and retry campaign
+			ee.log.Debug("campaigning to become leader")
+			err := election.Campaign(ee.cli.Ctx(), ee.conf.Name)
+			if err != nil {
+				switch {
+				case err == context.Canceled:
+					return
+				default:
+					continue
+				}
+			}
+			ee.log.Debug("won election to become leader")
+			// Resign as leader if you cannot start the leader funtions
+			if err := ee.startLeader(); err != nil {
```

Comment:
startLeader() underwent a lot of refactors as I developed this. Somewhere along the line, the errors that were present just got moved out. I'll have a look to see if I can get errors propagated back here.